lib/IRGen/GenThunk.cpp

```diff
@@ -0,0 +1,98 @@
+//===--- GenDecl.cpp - IR Generation for Method Dispatch Thunks -----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements IR generation for class and protocol method dispatch
+//  thunks, which are used in resilient builds to hide vtable and witness table
+//  offsets from clients.
+//
+//===----------------------------------------------------------------------===//
+
+#include "Callee.h"
+#include "Explosion.h"
+#include "GenDecl.h"
+#include "GenOpaque.h"
+#include "GenProto.h"
+#include "IRGenFunction.h"
+#include "IRGenModule.h"
+#include "ProtocolInfo.h"
+#include "Signature.h"
+#include "swift/IRGen/Linking.h"
+#include "llvm/IR/Function.h"
+
+using namespace swift;
+using namespace irgen;
+
+/// Find the entry point for a method dispatch thunk.
+llvm::Function *
+IRGenModule::getAddrOfDispatchThunk(SILDeclRef declRef,
+                                    ForDefinition_t forDefinition) {
+  LinkEntity entity = LinkEntity::forDispatchThunk(declRef);
+
+  llvm::Function *&entry = GlobalFuncs[entity];
+  if (entry) {
+    if (forDefinition) updateLinkageForDefinition(*this, entry, entity);
+    return entry;
+  }
+
+  auto &constantInfo = getSILModule().Types.getConstantInfo(declRef);
+  Signature signature = getSignature(constantInfo.SILFnType);
+  LinkInfo link = LinkInfo::get(*this, entity, forDefinition);
+
+  return createFunction(*this, link, signature);
+}
+
+static FunctionPointer lookupMethod(IRGenFunction &IGF,
+                                    SILDeclRef declRef) {
+  // Get the thunk type, which is the same as the type of the protocol
+  // requirement, and collect parameters.
+  auto &constantInfo = IGF.IGM.getSILModule().Types.getConstantInfo(declRef);
+  Signature signature = IGF.IGM.getSignature(constantInfo.SILFnType);
+
+  auto *decl = cast<AbstractFunctionDecl>(declRef.getDecl());
+  auto *dc = decl->getDeclContext();
+  auto *proto = cast<ProtocolDecl>(dc);
+
+  // Find the witness table.
+  llvm::Value *wtable = (IGF.CurFn->arg_end() - 1);
+
+  // Find the witness we're interested in.
+  auto &fnProtoInfo = IGF.IGM.getProtocolInfo(proto);
+  auto index = fnProtoInfo.getFunctionIndex(decl);
+  llvm::Value *witnessFnPtr =
+    emitInvariantLoadOfOpaqueWitness(IGF, wtable, index);
+
+  witnessFnPtr = IGF.Builder.CreateBitCast(witnessFnPtr,
+                                           signature.getType()->getPointerTo());
+
+  return FunctionPointer(witnessFnPtr, signature);
```

Comment:
Most of this is redundant with the second half of emitWitnessMethodValue, once it's computed the witness table pointer.  Please just extract that out as a separate function instead of duplicating it here.

Reply:
Done